core/src/test/java/kafka/testkit/TestKitNodes.java

```diff
@@ -59,73 +65,64 @@ public Builder setCombined(boolean combined) {
-            }
-
-            while (controllerNodeBuilders.size() > numControllerNodes) {
-                controllerNodeBuilders.pollFirstEntry();
-            }
-            while (controllerNodeBuilders.size() < numControllerNodes) {
-                int nextId = startControllerId();
-                if (!controllerNodeBuilders.isEmpty()) {
-                    nextId = controllerNodeBuilders.lastKey() + 1;
-                }
-                controllerNodeBuilders.put(nextId,
-                    new ControllerNode.Builder().
-                        setId(nextId));
-            }
+            this.numControllerNodes = numControllerNodes;
             return this;
         }
 
         public Builder setNumBrokerNodes(int numBrokerNodes) {
-            return setBrokerNodes(numBrokerNodes, 1);
+            this.numBrokerNodes = numBrokerNodes;
+            return this;
+        }
+
+        public Builder setNumDisksPerBroker(int numDisksPerBroker) {
+            this.numDisksPerBroker = numDisksPerBroker;
+            return this;
+        }
+
+        public Builder setPerBrokerProperties(Map<Integer, Map<String, String>> perBrokerProperties) {
+            this.perBrokerProperties = Collections.unmodifiableMap(
+                perBrokerProperties.entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, e -> Collections.unmodifiableMap(new HashMap<>(e.getValue())))));
+            return this;
         }
 
-        public Builder setBrokerNodes(int numBrokerNodes, int disksPerBroker) {
+        public TestKitNodes build() {
+            if (numControllerNodes < 0) {
+                throw new RuntimeException("Invalid negative value for numControllerNodes");
+            }
             if (numBrokerNodes < 0) {
                 throw new RuntimeException("Invalid negative value for numBrokerNodes");
             }
-            if (disksPerBroker <= 0) {
-                throw new RuntimeException("Invalid value for disksPerBroker");
-            }
-            while (brokerNodeBuilders.size() > numBrokerNodes) {
-                brokerNodeBuilders.pollFirstEntry();
+            if (numDisksPerBroker <= 0) {
+                throw new RuntimeException("Invalid value for numDisksPerBroker");
             }
-            while (brokerNodeBuilders.size() < numBrokerNodes) {
-                int nextId = startBrokerId();
-                if (!brokerNodeBuilders.isEmpty()) {
-                    nextId = brokerNodeBuilders.lastKey() + 1;
-                }
-                BrokerNode.Builder brokerNodeBuilder = new BrokerNode.Builder()
-                        .setId(nextId)
-                        .setNumLogDirectories(disksPerBroker);
-                brokerNodeBuilders.put(nextId, brokerNodeBuilder);
-            }
-            return this;
-        }
 
-        public TestKitNodes build() {
             String baseDirectory = TestUtils.tempDirectory().getAbsolutePath();
```

Comment:
We don't need to delete `baseDirectory` since `TestUtils.tempDirectory()` will delete the return folder when terminating.